orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -3131,3 +3142,199 @@ def test_debug_options_log_level_validation_fails_on_type_mismatch():
         _ = DebugOptions(log_level=log_level)
     assert f"Expected log_level of type LogLevel, got {type(log_level)}." in str(ex_info.value)
+
+@pytest.mark.parametrize("is_training,fallback_enabled,matching_policy", [(True, True, True), (False, True, True), (True, False, True), (False, False, True),
+                                                                          (True, True, False), (False, True, False), (True, False, False), (False, False, False)])
+def test_ortmodule_fallback_forward(is_training, fallback_enabled, matching_policy):
+    # is_training: True for torch.nn.Module training model, eval mode otherwise
+    # fallback_enabled: True results in PyTorch executing the forward graph instead of ORT backend
+    # matching_policy: True results in properly matching FALLBACK_UNSUPPORTED_DEVICE policy to ORTModuleDeviceException exception.
+    #   Otherwise, an incorrect policy (FALLBACK_UNSUPPORTED_INPUT) is used to verify that the fallback does not happen
+
+    from dataclasses import dataclass
+    @dataclass
+    class Point:
+        x: int
+        y: int
+
+    class UnsupportedInputModel(torch.nn.Module):
+        def __init__(self):
+            super(UnsupportedInputModel, self).__init__()
+        def forward(self, point):
+            return point.x * point.y
+
+    pt_model = UnsupportedInputModel()
+    ort_model = ORTModule(copy.deepcopy(pt_model))
+    inputs = Point(x=2,y=3)
+
+    policy = _fallback._FallbackPolicy.FALLBACK_FORCE_TORCH_FORWARD if matching_policy else _fallback._FallbackPolicy.FALLBACK_UNSUPPORTED_DEVICE
+
+    ort_model.train(is_training)
+    pt_model.train(is_training)
+    if fallback_enabled:
+        ort_model._torch_module._execution_manager(is_training=is_training)._fallback_manager._policy = policy
+        if matching_policy:
+            ort_out = ort_model(inputs)
+            pt_out = pt_model(inputs)
+            assert ort_out == pt_out
+        else:
+            with pytest.raises(_fallback.ORTModuleFallbackException) as type_error:
+                ort_model(inputs)
+            assert "ORTModule does not support the following model data type" in str(type_error.value)
+    else:
+        ort_model._torch_module._execution_manager(is_training=is_training)._fallback_manager._policy = _fallback._FallbackPolicy.FALLBACK_DISABLE
+        with pytest.raises(_fallback.ORTModuleFallbackException) as type_error:
+            ort_model(inputs)
+        assert "ORTModule does not support the following model data type" in str(type_error.value)
+
+
+@pytest.mark.parametrize("is_training,fallback_enabled,matching_policy", [(True, True, True), (False, True, True), (True, False, True), (False, False, True),
+                                                                          (True, True, False), (False, True, False), (True, False, False), (False, False, False)])
+def test_ortmodule_fallback_device__multiple(is_training, fallback_enabled, matching_policy):
```

Comment:
__ to _ ?

Reply:
Name is not great, but the idea is that `test_ortmodule_fallback_device` is a prefix for all fallback due to device issues. the `__XXXX` is a cue to the reason. There is one `mismatch` for device mismatch and another `multiple` for multiple device on the same model